flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -144,12 +145,340 @@ mlir::Value calculateTripCount(fir::FirOpBuilder &builder, mlir::Location loc,
+
+  mlir::Value outerLoopIV;
+  for (auto liveIn : outerLoopLiveIns) {
+    outerLiveInsSet.insert(getUnifiedLiveInView(liveIn));
+
+    // Keep track of the IV of the outerloop. See `isPerfectlyNested` for more
+    // info on the reason.
+    if (outerLoopIV == nullptr)
+      outerLoopIV = getUnifiedLiveInView(liveIn);
+  }
+
+  // For the 2 loops to be perfectly nested, either:
+  // * both would have exactly the same set of live-in values or,
+  // * the outer loop would have exactly 1 extra live-in value: the outer
+  //   loop's induction variable; this happens when the outer loop's IV is
+  //   *not* referenced in the nested loop.
+  bool isPerfectlyNested = [&]() {
+    if (outerLiveInsSet == nestedLiveInsSet)
+      return true;
+
+    if ((outerLiveInsSet.size() == nestedLiveIns.size() + 1) &&
+        !nestedLiveInsSet.contains(outerLoopIV))
+      return true;
+
+    return false;
+  }();
+
+  if (!isPerfectlyNested)
+    return mlir::failure();
+
+  return collectLoopNest(nestedUnorderedLoop, nestedLiveIns, loopNest);
+}
+
+/// Prepares the `fir.do_loop` nest to be easily mapped to OpenMP. In
+/// particular, this function would take this input IR:
+/// ```
+/// fir.do_loop %i_iv = %i_lb to %i_ub step %i_step unordered {
+///   fir.store %i_iv to %i#1 : !fir.ref<i32>
+///   %j_lb = arith.constant 1 : i32
+///   %j_ub = arith.constant 10 : i32
+///   %j_step = arith.constant 1 : index
+///
+///   fir.do_loop %j_iv = %j_lb to %j_ub step %j_step unordered {
+///     fir.store %j_iv to %j#1 : !fir.ref<i32>
+///     ...
+///   }
+/// }
+/// ```
+///
+/// into the following form (using generic op form since the result is
+/// technically an invalid `fir.do_loop` op:
+///
+/// ```
+/// "fir.do_loop"(%i_lb, %i_ub, %i_step) <{unordered}> ({
+/// ^bb0(%i_iv: index):
+///   %j_lb = "arith.constant"() <{value = 1 : i32}> : () -> i32
+///   %j_ub = "arith.constant"() <{value = 10 : i32}> : () -> i32
+///   %j_step = "arith.constant"() <{value = 1 : index}> : () -> index
+///
+///   "fir.do_loop"(%j_lb, %j_ub, %j_step) <{unordered}> ({
+///   ^bb0(%new_i_iv: index, %new_j_iv: index):
+///     "fir.store"(%new_i_iv, %i#1) : (i32, !fir.ref<i32>) -> ()
+///     "fir.store"(%new_j_iv, %j#1) : (i32, !fir.ref<i32>) -> ()
+///     ...
+///   })
+/// ```
+///
+/// What happened to the loop nest is the following:
+///
+/// * the innermost loop's entry block was updated from having one operand to
+///   having `n` operands where `n` is the number of loops in the nest,
+///
+/// * the outer loop(s)' ops that update the IVs were sank inside the innermost
+///   loop (see the `"fir.store"(%new_i_iv, %i#1)` op above),
+///
+/// * the innermost loop's entry block's arguments were mapped in order from the
+///   outermost to the innermost IV.
+///
+/// With this IR change, we can directly inline the innermost loop's region into
+/// the newly generated `omp.loop_nest` op.
+void preprocessLoopNest(mlir::ConversionPatternRewriter &rewriter,
```

Comment:
`preprocessLoopNest` is too generic for a name. `prepareLoopNest`? `canonicalizeLoopNest`?

How much effort would it be to avoid this step? Would prefer if we could avoid temporary IR changes.

Reply:
> Would prefer if we could avoid temporary IR changes.

If the conversion fails for any reason, the changes to the IR won't commited. If conversion succeeds, the whole loop nest will be erased. So no intermediate state of the loop will be seen in any case.

> `preprocessLoopNest` is too generic for a name. `prepareLoopNest`? `canonicalizeLoopNest`?

I agree it is too generic. `prepareLoopNest` is as generic though IMO. `canonicalizeLoopNest` alludes that there is some canonical form that we target which I think is not exactly true here. How about `moveLoopNestIVArgsToInnermostLoop`? A bit lengthy but captures what I am trying to do here.

Reply:
> canonicalizeLoopNest alludes that there is some canonical form that we target which I think is not exactly true here

A reason to avoid such temporary changes, but I won't insist.

> How about moveLoopNestIVArgsToInnermostLoop

Could be shortened to `sinkLoopIVArgs`. (sink == move into loop body)